components/event-sources/reconciler/httpsource/reconcile.go

```diff
@@ -0,0 +1,231 @@
+	messagingClient messagingclientv1alpha1.MessagingV1alpha1Interface
+
+	// URI resolver for sink destinations
+	sinkResolver *resolver.URIResolver
+}
+
+// Reconcile compares the actual state of a HTTPSource object referenced by key
+// with its desired state, and attempts to converge the two.
+func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
+	src, err := httpSourceByKey(key, r.httpsourceLister)
+	if err != nil {
+		return errors.Handle(err, ctx, "Failed to get object from local store")
+	}
+
+	currentKsvc, err := r.getOrCreateKnService(src)
+	if err != nil {
+		return err
+	}
+
+	currentCh, err := r.getOrCreateChannel(src)
+	if err != nil {
+		return err
+	}
+
+	desiredKsvc := r.makeKnService(src, currentKsvc)
+	currentKsvc, err = r.syncKnService(currentKsvc, desiredKsvc)
+	if err != nil {
+		return pkgerrors.Wrap(err, "failed to synchronize Knative Service")
+	}
+
+	return r.syncStatus(src, currentCh, currentKsvc)
+}
+
+// httpSourceByKey retrieves a HTTPSource object from a lister by ns/name key.
+func httpSourceByKey(key string, l sourceslistersv1alpha1.HTTPSourceLister) (*sourcesv1alpha1.HTTPSource, error) {
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return nil, controller.NewPermanentError(pkgerrors.Wrap(err, "invalid object key"))
+	}
+
+	src, err := l.HTTPSources(ns).Get(name)
+	switch {
+	case apierrors.IsNotFound(err):
+		return nil, errors.NewSkippable(pkgerrors.Wrap(err, "object no longer exists"))
+	case err != nil:
+		return nil, err
+	}
+
+	return src, nil
+}
+
+// getOrCreateKnService returns the existing Knative Service for a given
+// HTTPSource, or creates it if it is missing.
+func (r *Reconciler) getOrCreateKnService(src *sourcesv1alpha1.HTTPSource) (*servingv1alpha1.Service, error) {
+	ksvc, err := r.ksvcLister.Services(src.Namespace).Get(src.Name)
+	switch {
+	case apierrors.IsNotFound(err):
+		ksvc, err = r.servingClient.Services(src.Namespace).Create(r.makeKnService(src))
+		if err != nil {
+			return nil, pkgerrors.Wrap(err, "failed to create Knative Service")
+		}
+	case err != nil:
+		return nil, pkgerrors.Wrap(err, "failed to get Knative Service from cache")
+	}
+	return ksvc, nil
+}
+
+// getOrCreateChannel returns the existing Channel for a given HTTPSource, or
+// creates it if it is missing.
+func (r *Reconciler) getOrCreateChannel(src *sourcesv1alpha1.HTTPSource) (*messagingv1alpha1.Channel, error) {
+	ch, err := r.chLister.Channels(src.Namespace).Get(src.Name)
+	switch {
+	case apierrors.IsNotFound(err):
+		ch, err = r.messagingClient.Channels(src.Namespace).Create(r.makeChannel(src))
+		if err != nil {
+			return nil, pkgerrors.Wrap(err, "failed to create Channel")
+		}
+	case err != nil:
+		return nil, pkgerrors.Wrap(err, "failed to get Channel from cache")
+	}
+	return ch, nil
+}
+
+// makeKnService returns the desired Knative Service object for a given
+// HTTPSource. An optional Knative Service can be passed as parameter, in which
```

Comment:
the comment should somehow reflect that only 0 or 1 knative services can be passed in the variadic currrentKsvc.

Reply:
Isn't it clear from the phrase "An optional Knative Service can be passed as parameter"? It's singular, and there is the word optional. For an internal method that will do IMO.